### PR TITLE
Add jemalloc support

### LINKED ISSUE #182: Ruby with jemalloc
I would like to crate a ruby version compiled with jemalloc. Is this something the community wants is it encouraged to make an official contribution. If so I can try to create an official contribution.

### PR DESCRIPTION
**Pull request by marcotc**
Third time is the charm!
I had this PR open for a few weeks and realized there's work being done around `jemalloc`.

~~Should we add jemalloc only for Debian? If so, I'll just move my logic to the `template` file.~~

This PR adds `jemalloc` all Linux images: `debian` and `slim`.

~~I tried to make it generic, but I'm not sure we need this level of complexity here.~~

fixes #182, closes #190

### CODE REVIEW

**Comment by marcotc** [comment]
My initial intention was to create a new set of images, to avoid any backwards incompatibility, if any:

* 2.5.0-jemalloc
* 2.5.0-stretch-jemalloc
* 2.5.0-slim-jemalloc
* 2.5.0-slim-stretch-jemalloc

But it seems like there are talks about just adding `jemalloc` the default Debian images.

File: Dockerfile-debian.template [hunk 1]
```diff
@@ -25,8 +25,11 @@ RUN set -ex \
 		libssl1.0-dev \
 		ruby \
 	' \
+	&& runtimeDeps=' \
+	    libjemalloc-dev
+	' \
```
Comment (marcotc) [inline Dockerfile-debian.template]: I had to add it as a separate set of dependencies, because `buildDeps` get purged after the build is finished.

**Comment by ianks** [comment]
Is there a particular reason not to do this with alpine as well?

**Comment by marcotc** [comment]
@ianks not particularly. I'll give at shot at building alpine images.

**Comment by marcotc** [comment]
@ianks I just tried to compile Alpine with `jemalloc` and ended up getting stuck on this Ruby/Alpine/Jemalloc bug: https://bugs.ruby-lang.org/issues/13524

It seems like we are out of luck for the time being for Alpine+jemalloc.

**Comment by marcotc** [comment]
@n-rodriguez It seems like this has to do with how Ruby and Redis handle the pointer returned by [getcwd](https://linux.die.net/man/3/getcwd) differently:
* Ruby frees the returned pointer, as the Linux man page suggests. This can be seen in the code snippet provided by [shyouhei in the bug report](https://bugs.ruby-lang.org/issues/13524#note-7). [(Direct link to snippet)](https://github.com/shyouhei/docker-library/blob/master/%2313524/segv.c)
* Redis doesn't seem to free the returned pointer: https://github.com/antirez/redis/search?l=C&q=getcwd

File: 2.2/jessie/Dockerfile [hunk 2]
```diff
@@ -62,6 +66,9 @@ RUN set -ex \
 	&& gem install bundler --version "$BUNDLER_VERSION" --force \
 	&& rm -r /root/.gem/
 
+# TODO remove me before merging
+RUN ruby -r rbconfig -e "RbConfig::CONFIG['LIBS'].include?('jemalloc') ? puts('Ruby is compiled with jemalloc') : warn('JEMALLOC IS MISSING FROM RUBY')"
```
Comment (grosser) [inline 2.2/jessie/Dockerfile]: maybe nice to leave this as a sanity check with `RUN ruby -r rbconfig -e "abort 'jemalloc not enabled' unless RbConfig::CONFIG['LIBS'].include?('jemalloc')"`

**Comment by callumj** [comment]
Thanks for working on this @marcotc, we're currently ["patching"](https://github.com/hoteltonight/ruby) jemalloc support off the official Docker library using some [custom scripts](https://github.com/hoteltonight/ruby/blob/master/inject_jemalloc.rb) so this is a huge help by getting it official.

**Comment by ababich** [comment]
There is no issues in https://github.com/hoteltonight/ruby so may be I can comment here

As I understood this is published in Docker Hub as https://hub.docker.com/r/hoteltonight/ruby-jemalloc/ - is it possible to make relations between docker hub, github repo, etc. more transparent?

**Comment by cheerfulstoic** [comment]
I don't understand the issues that are going on, but it would be awesome to have jmalloc support built-in to our Ruby containers.  Is this stalled?

**Comment by yosifkit** [comment]
I think we were still uncertain whether this is the recommended or endorsed way to build ruby according to ruby upstream. We try to keep the images as close to upstream release recommendations as possible.

See also https://github.com/docker-library/ruby/pull/192#issuecomment-368612990.

**Comment by marcotc** [comment]
Would a PR with a new separate set of debian images using jemalloc be acceptable?
I can make those changes if that's the case.

**Comment by cheerfulstoic** [comment]
Also, would it be possible to specify an environment variable when building?  There seems to be a `--with-jemalloc` flag which could be used (though I imagine there would need to be at least one other conditional to install jemalloc beforehand

**Comment by ashleyhull-versent** [comment]
Hey @marcotc , #241 will hopefully fix the alpine blocker.

**Comment by marcotc** [comment]
Jemalloc is not "sticking" to Ruby 2.6rc as it does with older versions.
I don't see the library linked anymore, even though we built with it:

```sh
Step 11/17 : RUN ruby -r rbconfig -e "puts RbConfig::CONFIG['LIBS']"
 ---> Running in e0dc1e0e4115
-lm
```

Still investigating this.

**Comment by marcotc** [comment]
Now we have the scenario.
Jemalloc works fine for Ruby < 2.6 on Debian, and possibly Alpine 3.8, which would make the template file even more complex.

I'm wondering if we should just publish a separate set of images with the prefix `-jemalloc`, and not touch the "vanilla" images.

**Comment by tianon** [comment]
I wholeheartedly agree, and recommend that such work happen in a separate repo/image until Ruby upstream decides `jemalloc` is something they want to officially support _and_ recommend for general use. :+1:

**Comment by marcotc** [comment]
Just to clarify @tianon, are you saying `jemalloc` images should not be hosted under https://hub.docker.com/_/ruby/, not even if labelled appropriately?

**Comment by tianon** [comment]
I'm saying the docker-library team would prefer to not maintain them
unless/until they're officially recommended and supported by Ruby upstream.

I do see value in them existing, and recommend that someone create and
maintain them elsewhere, which will serve to help make them easier for
folks to consume and test, which will hopefully help convince upstream that
jemalloc is something they can rely on and should officially support or
even recommend outright.

Hopefully that helps clarify our position (and our hesitance to do anything
further with this PR at this time).

**Comment by marcotc** [comment]
I totally understand @tianon, thanks for clarifying.

**Comment by ashleyhull-versent** [comment]
Is there a forked image I can use? or are we not there yet?

**Comment by ababich** [comment]
@ashleyhull-versent I used hoteltonight image for some time but it is much easier to build your own based on official docker image, just take a look at 2-3 changes in Dockerfile and go ahead

**Comment by ashleyhull-versent** [comment]
Does anybody know who's actually at fault with this? Is there a ticket upstream with Ruby or Alpine?

**Comment by yosifkit** [comment]
Closing PR until a recommendation from Ruby upstream. Discussion can continue on #182.

See also https://github.com/docker-library/ruby/issues/182#issuecomment-456221587

**Pull request closed without merging**